sys-utils/libcmt/README.md

```diff
@@ -134,6 +134,124 @@ cast calldata-decode "Notice(bytes)" 0x`xxd -p -c0 "$1"` | (
+
+```
+from cffi import FFI
+ffi = FFI()
+
+def prune(file):
+    pruned = ""
+    ifmacro = False
+    for line in file.read().split("\n"):
+        if line.find("#endif") >= 0:
+            ifmacro = False
+            continue
+        if (line.find("#if") >= 0) and (line.find("#ifndef") < 0):
+            ifmacro = True
+            continue
+        if ifmacro:
+            continue
+
+        if line.find("#ifndef") >= 0:
+            continue
+        if line.find("#define") >= 0:
+            continue
+        if line.find("#include") >= 0:
+            continue
+
+        pruned = pruned + line + "\n"
+    return pruned
+
+
+import os
+headers = ""
+
+# Remove parts of the header cffi can't handle or has not implemented
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/buf.h")) as f:
+    headers = headers + prune(f)
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/abi.h")) as f:
+    headers = headers + prune(f)
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/keccak.h")) as f:
+    headers = headers + prune(f)
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/merkle.h")) as f:
+    headers = headers + prune(f)
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/io.h")) as f:
+    headers = headers + prune(f)
+with open(os.path.join(os.path.dirname(__file__), "libcmt-0.1.0/include/libcmt/rollup.h")) as f:
+    headers = headers + prune(f)
+
+# [1] Write the result to a file for inspection
+# with open(os.path.join(os.path.dirname(__file__), "cdef.h"), 'w') as f:
+#     f.write(headers)
+
+ffi.cdef(headers)
+ffi.set_source("pycmt",
+"""
+#include "libcmt/rollup.h"
+""",
+     include_dirs=["libcmt-0.1.0/include"],
+     library_dirs=["libcmt-0.1.0/lib"],
+     libraries=['cmt'])
+
+if __name__ == "__main__":
+    ffi.compile(verbose=True)
+```
+
+### Use
+
+With the module built, we can import it with python and use its functions.
+This example uses the raw bindings and just serves to ilustrate the process.
+Better yet would be to wrap these functions into a more "pythonic" API.
+
+`LD_LIBRARY_PATH=libcmt-0.1.0/lib/ python`
+
+```
+#!/usr/bin/env python
+
+# run this file only after building pycmt!
+import os
+from pycmt.lib import ffi, lib
+
+r = ffi.new("cmt_rollup_t[1]")
+assert(lib.cmt_rollup_init(r) == 0)
+
+address = ffi.new("uint8_t[20]", b"1000000000000")
+value   = ffi.new("uint8_t[32]", b"0000000000001")
+data    = b"hello world"
+index   = ffi.new("uint64_t *")
+print(os.strerror(-lib.cmt_rollup_emit_voucher(r,
+                            20, address,
+                            32, value,
+                            len(data), data, index)))
+
+ffi.gc(r, lib.cmt_rollup_fini)
+```
+
+Common errors such as the one below indicate that python couldn't find libcmt,
+make sure `LD_LIBRARY_PATH` points to the correct directory, or better yet, link
+against the static library.
```

Comment:
It does, for CFFI. It is even mentioned explicitly on the docs.
Python creates `.so` module incorporating the `.a` such that the module is the only thing required on the target.
```
...
Note this method works equally well with a static library piapprox.lib (Windows) or libpiapprox.a.
...
```

ref: https://cffi.readthedocs.io/en/latest/overview.html#id14

Reply:
Can Python actually build itself another shared library on top of a `.a` library. Python stuff, is that it?

Reply:
I see now. But still would need `LD_LIBRARY_PATH` to be correct to find this other `.so` despite statically linking , because it is actually is using a shared library as far as I understood, just "wrapped" over the static library one?

Reply:
Here is how it works:
`build` step:
- `ffi.cdefs` consumes the headers for the types we want to expose to python.
- `ffi.set_source` sets up compiler / linker flags, libraries and additional sources if needed.
  - we can say on `libraries` that we want the `.so` or `.a` to link against.
- `ffi.compile` will generate a `.c` file with all the bindings and then compile it into a `.so` python module.

If we link against an `.so`, the python module will need it when loading. Thus LD_LIBRARY_PATH is required if the lib is not on the system path.
```
ldd pycmt.cpython-311-x86_64-linux-gnu.so 
        linux-vdso.so.1 (0x00007ffe33dcd000)
        libcmt.so => not found
        libc.so.6 => /lib64/libc.so.6 (0x00007f7e334ca000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7e336c3000)
```

If we link against a `.a` then the python module can be loaded directly, without needing to setup LD_LIBRARY_PATH. It knows where to search when `import` is invoked, similar to how Lua does it.
```
ldd pycmt.cpython-311-x86_64-linux-gnu.so 
        linux-vdso.so.1 (0x00007ffe26234000)
        libc.so.6 => /lib64/libc.so.6 (0x00007ff352700000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ff3528ff000)

Reply:
Ok, got it :). Python stuff.